### PR TITLE
Issue / 8824 Extract New REST Controller from Core `Authentication` Class

### PR DESCRIPTION
**Pull request by 10upsimon**
## Summary

Addresses issue:

- #8824

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+		$authentication->register();
+
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/user/data/get-token' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 'valid-auth-token', $response->data['token'] );
+	}
+
+	private function grant_manage_options_permission() {
+		// Setup SiteKit.
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php:221]:
```suggestion
		// Setup Site Kit.
```

File: tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+		$this->grant_manage_options_permission();
+
+		$request  = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/user/data/get-token' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertFalse( $response->data['token'] );
+	}
+
+	/**
+	 * Tests presence of oauth token with correctly authenticated user.
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php:193]:
```suggestion
	 * Tests presence of OAuth token with correctly authenticated user.
```

File: tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		$authentication->register();
+
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/user/data/authentication' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertTrue( $response->data['authenticated'] );
+	}
+
+	/**
+	 * Tests presence of no oauth token with non-authenticated user.
```
Comment (tofumatt) [inline tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php:176]:
```suggestion
	 * Tests presence of no OAuth token with non-authenticated user.
```

**Review by tofumatt — COMMENTED** [review]
Looks good, just a few comment tweaks then it's good to go 👍🏻

**Review by tofumatt — APPROVED** [review]

File: tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+		$this->assertTrue( has_filter( 'googlesitekit_rest_routes' ) );
+		$this->assertTrue( has_filter( 'googlesitekit_apifetch_preload_paths' ) );
+	}
+
+	/**
+	 * Tests disconnection logic.
+	 *
+	 * @return void
+	 */
+	public function test_disconnect() {
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/Authentication/REST_Authentication_ControllerTest.php:76]: As a controller test, this should invoke the actual endpoint like the others. This test only deals with `Authentication` so this test should actually be in `AuthenticationTest`

**Pull request merged**